### ISSUE TITLE
`unnecessary_sort_by` rule suggests using standard library in `no_std` program

### ISSUE THREAD

**Issue by Wyatt-Stanke** [rust-lang/rust-clippy#11524]
### Summary

If you have a `vec.sort_by` that resembles:
```rust
vec.sort_by(|a, b|, b.cmp(a))
``` 
in a `#![no_std]` program, it will suggest 
```rust
vec.sort_by_key(|b| std::cmp::Reverse(b))
```
as a fix, but since the program is `no_std` it will cause an error.

### Reproducer

I tried this code:

```rust
#![no_std]
 
fn test() -> Vec<i32> {
    let mut vec = vec![1, 2, 3];
    vec.sort_by(|a, b| (b + 1).cmp(&(a + 1)));
    vec
}
```

I expected to see this happen:
Either no warning or a fix that wouldn't cause an error.

Instead, this happened:
It suggested the fix `vec.sort_by_key(|b| std::cmp::Reverse(b + 1));` which gave the error `failed to resolve: use of undeclared crate or module std` -- it could have used `core::cmp::Reverse` instead.

### Version

```text
rustc 1.74.0-nightly (7d9bce327 2023-09-16)
binary: rustc
commit-hash: 7d9bce327332fef22cf687b44de58f06d30daeb0
commit-date: 2023-09-16
host: x86_64-apple-darwin
release: 1.74.0-nightly
LLVM version: 17.0.0
```

### Additional Labels

@rustbot label +i-suggestion-causes-error

**Comment by Wyatt-Stanke** [comment]
> **Error**: Label i-suggestion-causes-error can only be set by Rust team members
> 
> Please file an issue on GitHub at [triagebot](https://github.com/rust-lang/triagebot) if there's a problem with this bot, or reach out on [#t-infra](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra) on Zulip.

Makes sense it's only for team members

**Comment by Centri3** [comment]
It should use the result of `clippy_utils::std_or_core`
